### ISSUE TITLE
JavaFX onAction in FXML file not wired

### ISSUE THREAD

**Issue by wehe64** [griffon/griffon#248]
Hello,
I'm trying to state my controller actions in my xfml file, like in
```
<GridPane fx:id="mainWindow" fx:controller="sample.FailcaseController" .... >
...
<Button fx:id="clickButton" onAction="#click" />

```
my controller is like 
```
ArtifactProviderFor(GriffonController)
class FailcaseController extends AbstractGriffonController {
...
    @FXML
    @ControllerAction
    @Threading(Threading.Policy.INSIDE_UITHREAD_ASYNC)
    public void click(ActionEvent evt=null) {
    ....
    }
```

and in my View (a .groovy) I have 
```
@ArtifactProviderFor(GriffonView)
class FailcaseView extends AbstractJavaFXGriffonView {
...
    void initUI() {
    ...
    GridPane root = loadFromFXML('sample/failsave.fxml')
    ...
    }

```

but I keep getting
`javafx.fxml.LoadException: Error resolving onAction='#click', either the event handler is not in the Namespace or there is an error in the script.`

I cannot find the problem - is wiring of controller actions from the fxml file supported?
Sources attached - and thanks for Griffon, I love it.
cheers, weHe
[failcase.zip](https://github.com/griffon/griffon/files/1241643/failcase.zip)

**Comment by aalmiray** [comment]
Aha, I see what's going on, there are two conventions being applied: JavaFX's fx:controller and Griffon's controller.

In standard JavaFX's FXML, the value set in `fx:controller` is used as the type where node injections (via `@FXML`) will take place; It's also used as the source for any event handlers, such as  `#click`.

In Griffon, when using the base class `AbstractJavaFXGriffonView` a different convention takes place. It really does not matter which value you define for `fx:controller` as the real instance is transparently updated to be the View, in your case `FailcaseView`. This means all `@FXML` injections occur on the View as expected. This also means all action handlers must be provided by the View. As this mixes logic in the View that (most of the times) belongs on the Controller, Griffon defines a different convention, see http://griffon-framework.org/guide/2.11.0/#_action_support

All this being said, my recommendation is to set the View class as the value for `fx:controller` as the IDE (IntelliJ in this case) can give you better code suggestion support (it can tell you if a widget with `fx:id` does not have a corresponding field annotated with `@FXML`; also if an event handler such as `#click` is not defined in the View too.

To summarize, the FXML file could look like this

	<GridPane fx:id="mainWindow" fx:controller="sample.FailcaseView">
	    <Button fx:id="clickButtonActionTarget"/>

or this too

	<GridPane fx:id="mainWindow" fx:controller="sample.FailcaseView">
	    <Button JavaFXUtils.griffonActionId="click"/>

Please note that there may be times when you want an UI only event handler (such as autoscrolling or element selection) in which case using a regular JavaFX onAction definition is OK. You would define such event handler in the View artifact.

I'll keep this issue open as I see there's missing documentation on the guide, where the distinction of JavaFX`s controller vs. Griffon's controller is stated clearly.

**Comment by aalmiray** [comment]
Additionally, you may want to have a look at the tutorials, specifically the JavaFX Views tutorial as it demonstrates different combinations http://griffon-framework.org/tutorials/4_javafx_views.html

**Comment by wehe64** [comment]
great! thank you very much, this made it totally clear. 
weHe

**Comment by wehe64** [comment]
BTW, there's a typo in http://griffon-framework.org/guide/2.11.0/#_action_support

> connectionActions(node, controller);

 should be

> connectActions(node, controller);

**Comment by wehe64** [comment]
and of corse, if you use the first method (fx:id="<controllerAction>ActionTarget") then you woun't be able to access the Button in the view for stuff like .disabled=... etc.

**Comment by aalmiray** [comment]
Typo fixed (thanks for spotting it!).

Regarding configuration of widget properties (such as disable), it turns out you have a couple of options to choose from:

If you follow the (fx:id="<controllerAction>ActionTarget")  option then you may inject the Button on the View as you would regularly do, such as

    @FXML private Button clickActionTarget;

You'll be able to tweak any properties as you have direct access to the widget.

If you follow the Griffon approach you may configure some action related properties in a declarative way, by setting properties inside the `griffon-app/i18n/messages.properties` file. The list of properties can be found at http://griffon-framework.org/guide/2.11.0/#_controllers_actionmanager
There's a ready made example at https://github.com/aalmiray/griffon-examples/tree/master/bindings

Or you can combine both, set some properties using the Griffon way and update others using the direct widget reference.

**Comment by wehe64** [comment]
Oh yeah, it had already dawned on me that I could name the Button like "...ActionTarget" and have it injected in my view that way... thanks anyway! I like to switch UI elements on / off etc depending on context...

And also thaks a great lot for that hint on the ActionManager, thats neat and really helpful stuff, man! Helps to keep the code lean - and I love the fact that it is bound to the i18n messages.

Having used Grails since 0.2 its a joy to see that paradigm available for JavaFX tasks - excellect work!
